src/lib/comms/include/sol-netctl.h

```diff
@@ -553,6 +696,81 @@ sol_netctl_find_service_by_name(const char *service_name)
 }
 
 /**
+ * @brief register a agent for network connection
+ *
+ * Register the agent for network connection.
+ * The register agent will inform the all information of network connection
+ * via the register callback struct.
+ * It includes the login input, the network connection failure and so on.
+ * The callback struct must be registered.
+ *
+ * @see struct sol_netctl_agent
+ *
+ * @param agent the agent Callback struct to be called when the related information is updated.
+ * @param data Add a user data per callback.
+ *
+ * @return 0 on success, -errno on failure.
+ */
+int
+sol_netctl_register_agent(const struct sol_netctl_agent *agent, const void *data);
+
+/**
+ * @brief report the error type to connection
+ *
+ * Report the error type for network connection.
+ * When the network connection failure, the user can select retry the connection or
+ * not retry it. The function can be used to select it.
+ * If retry connection is selected, the failure network connection will be tried to
+ * connect. If not retry connection is selected, the failure network connection will
+ * not be tried to connect.
+ * The connection failure information is informed via the agent callback.
+ * The agent must be registered before using sol_netctl_report_error.
+ *
+ * @see sol_netctl_register_agent
+ *
+ * @param service The service structure which the network connection is desired.
+ * @param type the selected network error type of the user.
+ *
+ * @return 0 on success, -errno on failure.
+ */
+int
+sol_netctl_report_error(struct sol_netctl_service *service,
+    const enum sol_netctl_agent_error_type type);
+
+/**
+ * @brief request the login input to connection
+ *
+ * Request the login input for network connection.
+ * When the login information is needed in the process of network connection,
+ * the funcation can be used to input the login information for network connection.
+ * The login input information is informed via the agent callback.
+ * The agent must be registered before using sol_netctl_request_input.
+ *
+ * @see sol_netctl_register_agent
+ *
+ * @param service The service structure which the network connection is desired.
+ * @param vector the vector of input types and inputs from the user.
+ *
+ * @return 0 on success, -errno on failure.
+ */
+int
+sol_netctl_request_input(struct sol_netctl_service *service,
+    const struct sol_ptr_vector *vector);
```

Comment:
same as before, try to use a vector of data, not vector-of-`ptr`-of-data. Rename the parameter to `inputs` and make clear in the doc the desired vector type.